feat: implementar creación de curso con TDD

morenavic opened the pull request:
- Se crearon los tests CursoTest, NivelTest y CrearCursoUseCaseTest.
- Se desarrolló el caso de uso CrearCursoUseCase basado en los tests.
- Se agregó el modelo Curso y el enum Nivel para representar niveles del curso.
- Se definieron excepciones para validar fecha, nivel y nombre.
- Se definieron las interfaces ICrearCursoInput y ICrearCursoRepositorio.
- Se actualizó el pom.xml para incluir dependencias necesarias.

morenavic committed:
feat: implementar creación de curso con TDD

- Se crearon los tests CursoTest, NivelTest y CrearCursoUseCaseTest.
- Se desarrolló el caso de uso CrearCursoUseCase basado en los tests.
- Se agregó el modelo Curso y el enum Nivel para representar niveles del curso.
- Se definieron excepciones para validar fecha, nivel y nombre.
- Se definieron las interfaces CrearCursoInput y ICrearCursoRepositorio.
- Se actualizó el pom.xml para incluir dependencias necesarias.

morenavic committed:
feat: creación de curso

- Se creó un test para validar que no se creen dos cursos con el mismo nombre.
- Se corrigieron tests en general.
- Nueva excepción CursoExisteException.

jruitti commented on core/src/main/java/curso/modelo/Curso.java:
```diff
@@ -0,0 +1,40 @@
+    private final LocalDateTime fechaCierreInscripcion;
+    private Nivel nivel;
+
+    public Curso(String nombre, LocalDateTime fechaCierreInscripcion, Nivel nivel) {
```
El constructor debe ser privado, así evitamos que sea invocado desde fuera

jruitti commented on core/src/main/java/curso/modelo/Curso.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    //Factory Method
+    public static Curso instancia(String nombre, LocalDateTime fecha_cierre_inscripcion, String nivelString) {
```
El uso del Enum para el Nivel es corecto, y su ventaja es que delimita los valores posibles. Pasarlo como String y validar ese string con el Enum me hace ruido. Lo podemos ver en clase el martes. Pero, a priori, este método debería recibir un Nivel como parámetro, no un String

jruitti commented on core/src/main/java/curso/usecase/CrearCursoUseCase.java:
```diff
@@ -0,0 +1,25 @@
+    @Override
+    public boolean crearCurso(Curso curso) {
+        if(iCrearCursoRepositorio.existe(curso.getNombre())){
+            throw new CursoExisteException("El curso ya existe.");
+        }else{
+            iCrearCursoRepositorio.guardar(curso);
+            return true;
```
Aquí hay una mejora, y una corrección.
Mejora: No es necesario el else cuando en el IF tenemos un throw (o return) como expresión. Es decir, con o sin el else, en este caso no se ejecutarán las instrucciones de las líneas 21 en adelante

Corrección: El método "guardar" del repositorio devuelve un boolean. Qué pasa si devuelve false?

jruitti commented on core/src/test/java/usecase/CrearCursoUseCaseTest.java:
```diff
@@ -0,0 +1,50 @@
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+public class CrearCursoUseCaseTest {
```
Faltaría un escenario, donde el curso no existe, pero la BD devuelve false

jruitti reviewed (commented):
Comentarios en cada línea correspondiente

morenavic committed:
feat: corrección

- Constructor Curso privado.
- El nivel se pasa como enum, se eliminan validaciones, test y excepción.
- Corrección del método crearCurso(), capturando el valor retornado por el método guardar().
- Nueva excepción GuardarCursoException.
- Nuevo test en CrearCursoUseCaseTest, el curso no existe y tampoco puede guardarse.
